ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,7 +579,107 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
     );
   }
 
+  private void verifyJoin(final Join root, final PlanNode left, final PlanNode right) {
+    if (left.getNodeOutputType() == DataSourceType.KSTREAM) {
+      if (right.getNodeOutputType() == DataSourceType.KTABLE) {
+
+        // stream-table join detected
+
+        if (root.getInfo().getType().equals(JoinType.OUTER)) {
+          throw new KsqlException("Invalid join type:"
+              + " full-outer join not supported for stream-table joins.");
+        }
+
+        // added outer check to make `multi-joins.json` pass
+        // the test uses expressions -> need to follow up on this issue
+        if (root.getInfo().getRightJoinExpression() instanceof ColumnReferenceExp) {
+          if (joinOnNonKeyAttribute(
+              ((ColumnReferenceExp) root.getInfo().getRightJoinExpression()),
+              right
+          )) {
+            throw new KsqlException("Invalid join condition:"
+                + " stream-table joins require to join on the table's primary key.");
+          }
+        }
+      }
+
+      // stream-stream join detected: nothing to verify
+
+    } else {
+      if (right.getNodeOutputType() == DataSourceType.KSTREAM) {
+        throw new KsqlException("Invalid join order:"
+            + " table-stream joins are not supported; only stream-table joins.");
+      }
+
+      // table-table join detected
+
+      if (joinOnNonKeyAttribute(
+            ((ColumnReferenceExp) root.getInfo().getLeftJoinExpression()),
+            left
+      )) {
+        throw new KsqlException("Invalid join condition:"
+            + " table-table joins require to join on the primary key of the left input table."
+        );
+      }
+
+      if (joinOnNonKeyAttribute(
+            ((ColumnReferenceExp) root.getInfo().getRightJoinExpression()),
+            right
+      )) {
+        throw new KsqlException("Invalid join condition:"
+            + " foreign-key table-table joins are not supported.");
+      }
+    }
+  }
+
+  private boolean joinOnNonKeyAttribute(final ColumnReferenceExp joinExpression,
+                                        final PlanNode node) {
+    final String joinAttributeName = joinExpression.getColumnName().text();
+    final List<DataSourceNode> dataSourceNodes = node.getSourceNodes().collect(Collectors.toList());
+    final String singleAttributeKeyName;
+    if (dataSourceNodes.size() == 1) {
```

Comment:
This is the binary-join case, for which `PlanNode` access a stream/table directly (ie, no nested join in PlanNode)

Reply:
Do we know that we've already performed the self-join check by this point? If not, is it possible that we are looking at a PlanNode representing a join, rather than a single-data source? A more reliable way to perform this check would be to check whether the PlanNode is a JoinNode or a DataSourceNode.

Reply:
The self-join check happens before IIRC -- we have test in place about it that still pass so it seems we don't break anything at this point. Can try out to check the node type as suggested.

Reply:
Did look into this, but we cannot do the `instanceof` check, because the `JoinNode` is "hidden" by the `ProJoinRepartitionNode` (and not exposed).

Reply:
Why can't we check for `instanceOf` `JoiningNode` instead of `JoinNode`?

Reply:
Well, we want to test the "child" of `node`, not `node` itself, ie, `((PreJoinRepartitionNode)node).joiningNode`.

We could of course expose it, but would it be any better?

Something like adding `PreJoinRepartitionNode#isInnerJoinNode()`.

Reply:
`node.getSource() instanceOf JoiningNode`?

Reply:
Hmm... we have only a `PlanNode` at hand, but `getSource()` is a method on `SingleSourcePlanNode` (that extends `PlanNode`).

Also, digging a little more, it seems we could have either a `JoinNode`, or `PreJoinRepartitionNode`, or `PreJoinProjectionNode` here and it's unclear how many layer to go down to find a `JoinNode`.

Maybe better to hold off for now and keep the simple check as-is. We can still refactor in a follow up if we think its required.

Reply:
About self-joins: I actually think they won't break this logic. We want to figure out if the join happens on the PK, and if there is self-join we may have just one data source, but it still seems to be correct to execute the `else` of this statement to figure out correctly if the join happens on the PK or not.

Reply:
A self-join can change the primary key of the result. For example, a table-table outer self-join would result in a new, system-generated primary key. If this output is then used in another join, that join should check that we are joining on the PK of the output of the join, not the original PK, right? (Or is that not supported, in which case this isn't a concern from a correctness point of view; it'd just be confusing since the method name would be a bit misleading in certain cases.)

Reply:
Good point! Damn you, outer-joins...